### PR TITLE
Joining groups

### PR DESCRIPTION
**Pull request by kmeyerhofer**
This PR's base is based on PR #13 

## Invitations Added
As a **group owner**, you can invite a user to join the group. Your invite can contain a message to the user. You can invite a user in two ways:
1. On the group page, you can send an invitation to a user's email. You must enter the email and a message.
2. On another user's profile page (`/users/:id`) it will show a drop down menu of the groups you are an owner of. Select the group, write a message and submit your invitation to that user.

You can only send one invitation to a user at a time. If the user declines the invitation, you cannot invite them again. 

If you have been invited to a group, your invitation will be at your `/dashboard` awaiting your Accept or Decline.

## Groups
* The `groups#destroy` functionality for a group owner works correctly deleting all `Item`s, `List`s, and finally the `Group` itself. 
* Groups now have `gift_due_date` date attributes. This is a required field during group creation and update. It will be used for business logic in the future.

## Users
* `User` email addresses are converted to downcase before being saved or accessed in the database. 
* `user#update` received a refactor: the current password is required to update any information for a user.

## Items
* Fixes update bug which wasn't saving `item#update` actions correctly.

#### `strong_password`
Adds the `strong_password` gem.
Users must use a strong password.
Password strength is tested in `spec/models/user_spec.rb`.

#### `database_cleaner`
For testing only.

## Gemfile 
Dependency upgrades.

## Database
* Migration created for the addition of `Invitations`.
* `db/seeds` updated to be more relevant for the new data.

## Tests!
* Adds Model and Feature tests for `users`, `groups`, `items`, `invitations`
* 95 examples are run with `rake` or `rspec`, one is currently not yet implemented.
* Most tests could be much more DRY.

### CODE REVIEW

**Comment by meyerhoferc** [comment]
@kmeyerhofer Please resolve merge conflicts. I think this are appearing because I just merged `wish-list` into `master`.

**Comment by meyerhoferc** [comment]
### Functional Testing Notes
#### Invitations
* Why are invitations on the user's show page? I don't see a way to find users other than knowing their ID to invite them to the group.
* In console, noticed issue with relationship between user + invitations. Can do `invitation.user` but not `user.invitations`. I think this is because the `Invitation` model has two relationships called `user` set.
* Invitations has show route but not action
* I am able to accept another user's invitation by manually typing in `/accept/:invitation_id`
* Unsure if this needs to be addressed: I was able to be added to a group twice when working through the console but not through the interface.
#### Groups
- works as advertised
#### Users
- Only issue is unable to update password, which I know you are working on separately.

File: app/controllers/dashboard_controller.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 class DashboardController < ApplicationController
   def show
-    if !current_user
+    if current_user
+      @user = current_user
+      @invitations = Invitation.where("receiver_id = ? AND accepted IS NULL", @user.id)
```
Comment (meyerhoferc) [inline app/controllers/dashboard_controller.rb]: Don't need to use `@user` here. Can use `current_user` in query for invitations and in the view (as it is a helper method)

Comment (meyerhoferc) [reply app/controllers/dashboard_controller.rb]: could use AR relationships with invitations for user (sent / received)

File: app/controllers/groups_controller.rb [hunk 2]
```diff
@@ -2,7 +2,7 @@ class GroupsController < ApplicationController
   before_action :set_group, only: [:show, :edit, :update, :destroy]
   before_action -> { unauthorized_user(@group.owner) }, only: [:update, :edit, :destroy]
 
-  def index
+  def index # Delete for the future?
     @groups = Group.all
   end
```
Comment (meyerhoferc) [inline app/controllers/groups_controller.rb]: why can everyone see this?

File: app/controllers/groups_controller.rb [hunk 3]
```diff
@@ -36,13 +40,28 @@ def update
...
     else
-      flash[:warning] = 'An error occurred, please try again.'
+      flash[:warning] = 'Please enter valid information.'
       redirect_to edit_group_path(@group)
     end
   end
 
   def destroy
-    # Destroy not working. Delete lists associated with group first.
+    @lists = List.where("group_id = :group_id", {group_id: @group.id})
```
Comment (meyerhoferc) [inline app/controllers/groups_controller.rb]: showing you alternative: `List.where(group_id: @group.id)`

File: app/controllers/invitations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    @invitation.sender_id = current_user.id
+    @invitation.receiver_id = params[:user_id]
+    @invitation.save
+    flash[:notice] = 'Invitation sent.'
+    redirect_to user_path(@invitation.receiver_id)
+  end
+
+  def accept
+    invitation = Invitation.find(params[:id])
+    user = User.find(invitation.receiver_id)
```
Comment (meyerhoferc) [inline app/controllers/invitations_controller.rb]: user = `invitation.receiver`

File: app/controllers/invitations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      end
+    else
+      flash[:warning] = 'Action is unauthorized.'
+      redirect_to root_path
+    end
+  end
+
+  def decline
+    invitation = Invitation.find(params[:id])
+    user = User.find(invitation.receiver_id)
```
Comment (meyerhoferc) [inline app/controllers/invitations_controller.rb]: `user = invitation.receiver`

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -52,22 +61,44 @@ def update
...
-      return params.require(:user).permit(:email)
+      downcase_email_param!
+      return params.require(:user).permit(:email, :password)
     when 'Password'
       return params.require(:user).permit(:password, :password_confirmation)
     end
   end
 
+  def invitable_groups
+    Group.where("owner_id = ?", current_user.id).select do |group|
```
Comment (meyerhoferc) [inline app/controllers/users_controller.rb]: needs to eventually be pushed down to model level with tests

File: app/models/user.rb [hunk 7]
```diff
@@ -2,9 +2,14 @@ class User < ApplicationRecord
   has_many :user_groups
   has_many :groups, through: :user_groups
   has_many :lists
-  has_secure_password
+  has_many :received, foreign_key: :receiver_id, class_name: 'Invitation'
+  has_many :sent, foreign_key: :sender_id, class_name: 'Invitation'
   validates_uniqueness_of :email
+  validates_uniqueness_of :id, scope: [:user_groups, :groups]
```
Comment (meyerhoferc) [inline app/models/user.rb]: check if this is needed given line 4 of `UserGroup` model validations

File: app/views/groups/_form.html.erb [hunk 8]
```diff
@@ -7,6 +7,10 @@
     <%= f.label :description %>
     <%= f.text_area :description %>
   </div>
+  <div>
+    <%= f.label :gift_due_date, 'Gift Due Date (YYYY/MM/DD)' %>
+    <%= f.text_field :gift_due_date %>
```
Comment (meyerhoferc) [inline app/views/groups/_form.html.erb]: there is a date field for rails form helper

File: app/controllers/dashboard_controller.rb (hunk 1 above)
Comment (kmeyerhofer) [reply app/controllers/dashboard_controller.rb]: Fixed. The query on line 5 has been moved to the `user` model, accessed through `user.outstanding_invitations`.

File: app/controllers/groups_controller.rb (hunk 2 above)
Comment (kmeyerhofer) [reply app/controllers/groups_controller.rb]: Mostly debugging purposes to view all the groups at once. I will remove it in this branch.

File: app/controllers/groups_controller.rb (hunk 3 above)
Comment (kmeyerhofer) [reply app/controllers/groups_controller.rb]: Much simpler, I'll use it.

File: app/models/user.rb (hunk 7 above)
Comment (kmeyerhofer) [reply app/models/user.rb]: Removed.

File: app/views/groups/_form.html.erb (hunk 8 above)
Comment (kmeyerhofer) [reply app/views/groups/_form.html.erb]: Feature added. (Pretty cool too!)

View logic added setting the minimum date allowed for entry.

**Pull request merged**